### PR TITLE
Finalize core22 transition (5.0-edge)

### PR DESCRIPTION
**Pull request by simondeziel**
This PR contains all the commits I deemed appropriate to cherry-pick from `latest-edge` right before it moved from `core22` to `core24` (so at commit e6c25ef57a4ca7e4217421fb453b1a31b8bbf2d9).

Here's what was not cherry-picked:

* apparmor `unconfined` changes
* uefi variable edition
* `edk2` commit dadf7717 `edk2: Cleanup build flags and switch to gcc5` as that caused issue with secure-boot cert injection
* `edk2` refreshed part definition/`override-build` section
* `CSM` support
* `minio` support
* `nasm` man page patch refresh (since `nasm` version was not bumped)
* `nvme` support
* `qemu` change to Ubuntu/LP source
* `apparmor` part removal
* `zfs-0.6`, `zfs-0.7` and `zfs-2.0` part removal
* parts versions updates
* `ceph` removal from `armhf`
* `lxc-to-lxd` removal
* `lxd-benchmark` removal
* `lxd-migrate` removal
* `nano` part removal
* `vim` part removal
* `shiftfs` removal
* `upgrade-bridge` script removal
* `XDG_*` variable removal
* anything from bots
* `LXD_OVMF_PATH` replacement by `LXD_QEMU_FW_PATH` as `stable-5.0` still uses the former
* `/var/lib/lxd/unix.socket` handling
* LXD-UI remains disabled
* `microceph` plug
* `microovn` plug
* license change
* LXD builtin documentation assets

I also tweaked the `zfs-0-6` part to no longer go a `git cherry-pick` + full clone but instead use (another) `.patch` file.

### CODE REVIEW

**Comment by tomponline** [comment]
@simondeziel please can you update the PR description to help me understand the methodology you used here to select the commits - was it primarily just doing a diff between 5.21-edge and 5.0-edge and then looking for the associated commit?

What I'm trying to figure out is what to look for in terms of risk areas/large changes. Thanks

**Comment by simondeziel** [comment]
Locally built snap:

```
Generated snap metadata                                                                                                                  
'--enable-manifest' is deprecated, and will be removed in core24.                                                                        
Version has been set to '0+git.2220cbf4'                                                                                                 
Generated snap manifest                                                                                                                  
Lint warnings:                                                                                                                           
- library: libapparmor.so.1: unused library 'lib/libapparmor.so.1.17.1'. (https://snapcraft.io/docs/linters-library)                     
- library: libnvidia-container-go.so.1: unused library 'lib/libnvidia-container-go.so.1.14.1'. (https://snapcraft.io/docs/linters-library)                                                                                                                                         
- library: libatomic.so.1: unused library 'lib/x86_64-linux-gnu/libatomic.so.1.2.0'. (https://snapcraft.io/docs/linters-library)         
- library: libjson-glib-1.0.so.0: unused library 'lib/x86_64-linux-gnu/libjson-glib-1.0.so.0.600.6'. (https://snapcraft.io/docs/linters-library)                                                                                                                                   
- library: libusbredirhost.so.1: unused library 'lib/x86_64-linux-gnu/libusbredirhost.so.1.0.2'. (https://snapcraft.io/docs/linters-library)                                                                                                                                       
- library: libnvpair.so.1: unused library 'zfs-0.6/lib/libnvpair.so.1.0.1'. (https://snapcraft.io/docs/linters-library)                  
- library: libuutil.so.1: unused library 'zfs-0.6/lib/libuutil.so.1.0.1'. (https://snapcraft.io/docs/linters-library)                    
- library: libzfs.so.2: unused library 'zfs-0.6/lib/libzfs.so.2.0.0'. (https://snapcraft.io/docs/linters-library)                        
- library: libzfs_core.so.1: unused library 'zfs-0.6/lib/libzfs_core.so.1.0.0'. (https://snapcraft.io/docs/linters-library)              
- library: libzpool.so.2: unused library 'zfs-0.6/lib/libzpool.so.2.0.0'. (https://snapcraft.io/docs/linters-library)                    
- library: libnvpair.so.1: unused library 'zfs-0.8/lib/libnvpair.so.1.0.1'. (https://snapcraft.io/docs/linters-library)                  
- library: libuutil.so.1: unused library 'zfs-0.8/lib/libuutil.so.1.0.1'. (https://snapcraft.io/docs/linters-library)                    
- library: libzfs.so.2: unused library 'zfs-0.8/lib/libzfs.so.2.0.0'. (https://snapcraft.io/docs/linters-library)                        
- library: libzfs_core.so.1: unused library 'zfs-0.8/lib/libzfs_core.so.1.0.0'. (https://snapcraft.io/docs/linters-library)              
- library: libzpool.so.2: unused library 'zfs-0.8/lib/libzpool.so.2.0.0'. (https://snapcraft.io/docs/linters-library)                    
- library: libzfsbootenv.so.1: unused library 'zfs-2.0/lib/libzfsbootenv.so.1.0.0'. (https://snapcraft.io/docs/linters-library)          
- library: libzpool.so.4: unused library 'zfs-2.0/lib/libzpool.so.4.0.0'. (https://snapcraft.io/docs/linters-library)                    
- library: libnvpair.so.3: unused library 'zfs-2.1/lib/libnvpair.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                  
- library: libuutil.so.3: unused library 'zfs-2.1/lib/libuutil.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                    
- library: libzfs.so.4: unused library 'zfs-2.1/lib/libzfs.so.4.1.0'. (https://snapcraft.io/docs/linters-library)                        
- library: libzfs_core.so.3: unused library 'zfs-2.1/lib/libzfs_core.so.3.0.0'. (https://snapcraft.io/docs/linters-library)              
- library: libzfsbootenv.so.1: unused library 'zfs-2.1/lib/libzfsbootenv.so.1.0.0'. (https://snapcraft.io/docs/linters-library)          
- library: libzpool.so.5: unused library 'zfs-2.1/lib/libzpool.so.5.0.0'. (https://snapcraft.io/docs/linters-library)                    
- library: libnvpair.so.3: unused library 'zfs-2.2/lib/libnvpair.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                  
- library: libuutil.so.3: unused library 'zfs-2.2/lib/libuutil.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                    
- library: libzfs.so.4: unused library 'zfs-2.2/lib/libzfs.so.4.1.0'. (https://snapcraft.io/docs/linters-library)                        
- library: libzfs_core.so.3: unused library 'zfs-2.2/lib/libzfs_core.so.3.0.0'. (https://snapcraft.io/docs/linters-library)              
- library: libzfsbootenv.so.1: unused library 'zfs-2.2/lib/libzfsbootenv.so.1.0.0'. (https://snapcraft.io/docs/linters-library)          
- library: libzpool.so.5: unused library 'zfs-2.2/lib/libzpool.so.5.0.0'. (https://snapcraft.io/docs/linters-library)                    
Version has been set to '0+git.2220cbf4'                                                                                                 
Created snap package lxd_0+git.2220cbf4_amd64.snap 
```

**Comment by simondeziel** [comment]
This snap was tested in local VMs using 22.04 and 24.04. A subset of the `lxd-ci` suite was ran in each. Minimal manual tests were also done in a 20.04 VM.

**Comment by tomponline** [comment]
> ceph removal from armhf

@simondeziel i dont think we can do this, this was supported in 5.0 and is still available in core22 right?

**Comment by tomponline** [comment]
> lxc-to-lxd removal
> lxd-benchmark removal
> lxd-migrate removal

These need to stay in - we are just changing the base and bringing in updates, not taking away functionality.

**Comment by tomponline** [comment]
> nano part removal
> vim part removal
> shiftfs removal
> upgrade-bridge script removal

These need to stay in, as it is a late LTS point release so support/behaviours should remain as stable as possible.

**Comment by simondeziel** [comment]
@tomponline yep, was easy to miss but the list in the PR description is for thing I did **not** cherry-pick :)

**Comment by tomponline** [comment]
Apologies I misread that you had *not* cherry-picked that list :)

File: .github/workflows/builds.yml [hunk 1]
```diff
@@ -70,6 +70,11 @@ jobs:
           localRev=$(git rev-parse HEAD)
           go install github.com/canonical/lxd-ci/lxd-snapcraft@latest
           git clone -b "${TARGET}" git+ssh://lxdbot@git.launchpad.net/~canonical-lxd/lxd ~/lxd-pkg-snap-lp
+          ver=($(lxd-snapcraft -get-version -file ~/lxd-pkg-snap-lp/snapcraft.yaml))
           rsync -a --exclude .git --delete . ~/lxd-pkg-snap-lp/
           cd ~/lxd-pkg-snap-lp
-          git diff
+          lxd-snapcraft -set-version "${ver[0]}" -set-source-commit "${ver[1]}"
```
Comment (tomponline) [inline .github/workflows/builds.yml]: do we not already have this? How are 5.0 snap builds working now?

File: snapcraft/commands/daemon.start [hunk 2]
```diff
@@ -463,6 +463,10 @@ if [ "${ui_enable:-"false"}" = "true" ]; then
     export LXD_UI="${SNAP_CURRENT}/share/lxd-ui/"
 fi
 
+# Setup documentation
```
Comment (tomponline) [inline snapcraft/commands/daemon.start]: Lets not add the docs to the snap for this release, that is outside of the focus for the LTS point release, and will grow the snap.

**Review by tomponline — COMMENTED** [review]
Did you add the docs to the snap because of the UI support?

File: snapcraft.yaml [hunk 3]
```diff
@@ -1623,6 +1614,7 @@ parts:
       # XXX: remove unneeded files/directories
       rm -rf "${SNAPCRAFT_PRIME}/usr/local/"
 
+<<<<<<< HEAD
```
Comment (tomponline) [inline snapcraft.yaml]: is this a left over conflict resolution?

File: snapcraft.yaml [hunk 4]
```diff
@@ -1170,18 +1186,20 @@ parts:
...
+      mv "${CRAFT_PART_INSTALL}.tmp/sbin/zpool" "${CRAFT_PART_INSTALL}/zfs-${ZFS_VER}/bin/"
+      mv "${CRAFT_PART_INSTALL}.tmp/lib/udev/zvol_id" "${CRAFT_PART_INSTALL}/zfs-${ZFS_VER}/bin/"
+      mv "${CRAFT_PART_INSTALL}.tmp/lib/"*so* "${CRAFT_PART_INSTALL}/zfs-${ZFS_VER}/lib/"
+      rm -Rf "${CRAFT_PART_INSTALL}.tmp"
 
   zfs-2-2:
     source: https://github.com/openzfs/zfs
     source-type: git
-    source-tag: zfs-2.2.2
+    source-tag: zfs-2.2.2-rc4
```
Comment (tomponline) [inline snapcraft.yaml]: This version change seems unexpected and not related to commit message.

File: snapcraft.yaml [hunk 5]
```diff
@@ -1058,17 +1069,18 @@ parts:
...
-      mv "${SNAPCRAFT_PART_INSTALL}" "${SNAPCRAFT_PART_INSTALL}.tmp"
-      mkdir -p "${SNAPCRAFT_PART_INSTALL}/zfs-0.7/bin" "${SNAPCRAFT_PART_INSTALL}/zfs-0.7/lib"
-      mv "${SNAPCRAFT_PART_INSTALL}.tmp/sbin/zfs" "${SNAPCRAFT_PART_INSTALL}/zfs-0.7/bin/"
-      mv "${SNAPCRAFT_PART_INSTALL}.tmp/sbin/zpool" "${SNAPCRAFT_PART_INSTALL}/zfs-0.7/bin/"
-      mv "${SNAPCRAFT_PART_INSTALL}.tmp/lib/udev/zvol_id" "${SNAPCRAFT_PART_INSTALL}/zfs-0.7/bin/"
-      mv "${SNAPCRAFT_PART_INSTALL}.tmp/lib/"*so* "${SNAPCRAFT_PART_INSTALL}/zfs-0.7/lib/"
-      rm -Rf "${SNAPCRAFT_PART_INSTALL}.tmp"
+      ZFS_VER="0.7"
+
+      mv "${CRAFT_PART_INSTALL}" "${CRAFT_PART_INSTALL}.tmp"
```
Comment (tomponline) [inline snapcraft.yaml:1089]: these changes dont appear related to commit message

File: snapcraft.yaml [hunk 6]
```diff
@@ -1633,32 +1633,24 @@ parts:
       set -x
 
       # XXX: remove unneeded files/directories
-      rm -rf "${SNAPCRAFT_PRIME}/usr/local/"
-
-<<<<<<< HEAD
```
Comment (tomponline) [inline snapcraft.yaml]: this looks like it fixed the issue from the earlier commit

File: snapcraft/commands/daemon.start [hunk 7]
```diff
@@ -442,20 +442,6 @@ if [ "$(stat -c '%u' /proc)" = 0 ]; then
             echo 1 > /proc/sys/kernel/unprivileged_userns_clone || true
         fi
     fi
-
-    if [ -e /proc/sys/kernel/apparmor_restrict_unprivileged_userns ]; then
```
Comment (tomponline) [inline snapcraft/commands/daemon.start:439]: do we need this commit, they should remain globally disabled in my view for Noble

File: snapcraft.yaml [hunk 8]
```diff
@@ -919,6 +914,18 @@ parts:
       - bin/tar2sqfs
       - lib/libsquashfs.so*
 
+  vim:
```
Comment (tomponline) [inline snapcraft.yaml]: adding vim does not appear to be related to commit message

File: lxd-migrate/lxd.go [hunk 9]
```diff
@@ -218,10 +218,6 @@ func (d *lxdDaemon) reload() error {
 		return systemdCtl("reload", "snap.lxd.daemon.service")
 	}
 
-	if osInit() == "upstart" {
```
Comment (tomponline) [inline lxd-migrate/lxd.go]: lets just leave this change out as we dont want to be removing stuff from the LTS

File: snapcraft.yaml [hunk 10]
```diff
@@ -1668,6 +1668,13 @@ parts:
         -not -path "${CRAFT_PRIME}/lib/liblxcfs.so" \
         -exec strip -s {} +
 
+      if [ "$(uname -m)" != "riscv64" ]; then
```
Comment (tomponline) [inline snapcraft.yaml]: this looks like a re-order but commit message seems different

File: lxd-migrate/main.go [hunk 11]
```diff
@@ -313,9 +313,9 @@ func askBool(question string, default_ string) bool {
 		if input == "" {
 			input = default_
 		}
-		if shared.StringInSlice(strings.ToLower(input), []string{"yes", "y"}) {
+		if shared.ValueInSlice(strings.ToLower(input), []string{"yes", "y"}) {
```
Comment (tomponline) [inline lxd-migrate/main.go:316]: does this build using 5.0's go version?

File: snapcraft.yaml [hunk 12]
```diff
@@ -1630,6 +1634,10 @@ parts:
       rm -rf "${CRAFT_PRIME}/lib/systemd/"
       rm -rf "${CRAFT_PRIME}/lib/udev/"
       rm -rf "${CRAFT_PRIME}/usr/local/"
+<<<<<<< HEAD
+=======
+      rm -rf "${CRAFT_PRIME}/usr/share/"
+>>>>>>> e436b6f2 (snapcraft: don't prime anything from the nasm part)
```
Comment (tomponline) [inline snapcraft.yaml]: this appeared now as part of a conflict resolution left over

File: snapcraft.yaml [hunk 13]
```diff
@@ -1634,10 +1634,7 @@ parts:
       rm -rf "${CRAFT_PRIME}/lib/systemd/"
       rm -rf "${CRAFT_PRIME}/lib/udev/"
       rm -rf "${CRAFT_PRIME}/usr/local/"
-<<<<<<< HEAD
-=======
       rm -rf "${CRAFT_PRIME}/usr/share/"
->>>>>>> e436b6f2 (snapcraft: don't prime anything from the nasm part)
```
Comment (tomponline) [inline snapcraft.yaml]: would be better to not add them in the first place rather than remove them via later commit

**Review by tomponline — CHANGES_REQUESTED** [review]
Overall good thanks.

I think the go source changes for Go programs inside the snap are fine in retrospect.

There's a few places where the commit messages dont match the changes and there are some left over cherry-pick conflict resolution lines, so think the rebasing might have gone awry at some point.

Also not sure about adding the embedded docs at this stage in the LTS lifecycle.

File: lxd-migrate/main.go (hunk 11 above)
Comment (simondeziel) [reply lxd-migrate/main.go:316]: Since we are building 5.0 with the latest Go version, I assumed it was fair game to rely on modern Go version for bits shipped by the snap. Ah, you later agreed with that decision, good :)

File: snapcraft.yaml (hunk 8 above)
Comment (simondeziel) [reply snapcraft.yaml]: It was a mistake to reintroduce a `vim` part as that was already dropped from `5.0-edge` by https://github.com/canonical/lxd-pkg-snap/pull/251

**Comment by simondeziel** [comment]
Locally built snap:
```
Generated snap metadata                                                                                                                                                                                                                                                            
'--enable-manifest' is deprecated, and will be removed in core24.                                                                                                                                                                                                                  
Version has been set to '0+git.0f0540b4'                                                                                                                                                                                                                                           
Generated snap manifest                                                                                                                                                                                                                                                            
Lint warnings:                                                                                                                                                                                                                                                                     
- library: libapparmor.so.1: unused library 'lib/libapparmor.so.1.17.1'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                               
- library: libnvidia-container-go.so.1: unused library 'lib/libnvidia-container-go.so.1.14.1'. (https://snapcraft.io/docs/linters-library)                                                                                                                                         
- library: libatomic.so.1: unused library 'lib/x86_64-linux-gnu/libatomic.so.1.2.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                   
- library: libjson-glib-1.0.so.0: unused library 'lib/x86_64-linux-gnu/libjson-glib-1.0.so.0.600.6'. (https://snapcraft.io/docs/linters-library)                                                                                                                                   
- library: libusbredirhost.so.1: unused library 'lib/x86_64-linux-gnu/libusbredirhost.so.1.0.2'. (https://snapcraft.io/docs/linters-library)                                                                                                                                       
- library: libnvpair.so.1: unused library 'zfs-0.6/lib/libnvpair.so.1.0.1'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                            
- library: libuutil.so.1: unused library 'zfs-0.6/lib/libuutil.so.1.0.1'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
- library: libzfs.so.2: unused library 'zfs-0.6/lib/libzfs.so.2.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                                  
- library: libzfs_core.so.1: unused library 'zfs-0.6/lib/libzfs_core.so.1.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                        
- library: libzpool.so.2: unused library 'zfs-0.6/lib/libzpool.so.2.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
- library: libnvpair.so.1: unused library 'zfs-0.8/lib/libnvpair.so.1.0.1'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                            
- library: libuutil.so.1: unused library 'zfs-0.8/lib/libuutil.so.1.0.1'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
- library: libzfs.so.2: unused library 'zfs-0.8/lib/libzfs.so.2.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                                  
- library: libzfs_core.so.1: unused library 'zfs-0.8/lib/libzfs_core.so.1.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                        
- library: libzpool.so.2: unused library 'zfs-0.8/lib/libzpool.so.2.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
- library: libzfsbootenv.so.1: unused library 'zfs-2.0/lib/libzfsbootenv.so.1.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                    
- library: libzpool.so.4: unused library 'zfs-2.0/lib/libzpool.so.4.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
- library: libnvpair.so.3: unused library 'zfs-2.1/lib/libnvpair.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                            
- library: libuutil.so.3: unused library 'zfs-2.1/lib/libuutil.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
- library: libzfs.so.4: unused library 'zfs-2.1/lib/libzfs.so.4.1.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                                  
- library: libzfs_core.so.3: unused library 'zfs-2.1/lib/libzfs_core.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                        
- library: libzfsbootenv.so.1: unused library 'zfs-2.1/lib/libzfsbootenv.so.1.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                    
- library: libzpool.so.5: unused library 'zfs-2.1/lib/libzpool.so.5.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
- library: libnvpair.so.3: unused library 'zfs-2.2/lib/libnvpair.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                            
- library: libuutil.so.3: unused library 'zfs-2.2/lib/libuutil.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
- library: libzfs.so.4: unused library 'zfs-2.2/lib/libzfs.so.4.1.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                                  
- library: libzfs_core.so.3: unused library 'zfs-2.2/lib/libzfs_core.so.3.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                        
- library: libzfsbootenv.so.1: unused library 'zfs-2.2/lib/libzfsbootenv.so.1.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                    
- library: libzpool.so.5: unused library 'zfs-2.2/lib/libzpool.so.5.0.0'. (https://snapcraft.io/docs/linters-library)                                                                                                                                                              
Version has been set to '0+git.0f0540b4'                                                                                                                                                                                                                                           
Created snap package lxd_0+git.0f0540b4_amd64.snap
```

And manual tests went well in a 22.04 VM with the 5.15 `-kvm` kernel and the 6.5 `-virtual-hwe-22.04` kernel as well. With those 2, ZFS 2.1 and 2.2 were tested.

Also tested in a 20.04 VM with the 5.4 `-kvm` kernel and using ZFS 0.8.

**Comment by simondeziel** [comment]
And I just found a regression on 18.04:

```
root@u1:~# snap install --dangerous /dev/shm/lxd_0+git.0f0540b4_amd64.snap 
2024-07-19T05:25:02Z INFO Waiting for automatic snapd restart...
error: cannot perform the following tasks:
- Run configure hook of "lxd" snap if present (run hook "configure": 
-----
cat: /proc/self/attr/current: Permission denied
/snap/lxd/x1/snap/hooks/configure: 6: exec: aa-exec: Permission denied
-----)
root@u1:~# ll /proc/self/attr/current
-rw-rw-rw- 1 root root 0 Jul 19 05:25 /proc/self/attr/current
root@u1:~# lsb_release -rc
Release:	18.04
Codename:	bionic
root@u1:~# uname -a
Linux u1 4.15.0-212-generic #223-Ubuntu SMP Tue May 23 13:09:22 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by simondeziel** [comment]
At first glance, this doesn't seem to be a regression on LXD as `git diff 5.0-edge.. -- snapcraft/hooks/configure` shows nothing changed in how Apparmor is escaped.

Maybe a snapd confinement change?

**Comment by simondeziel** [comment]
How many times will I fall for this one? `--dangerous` install require a prior official snap to be installed first:

```
root@u1:~# snap install --dangerous /dev/shm/lxd_0+git.0f0540b4_amd64.snap 
error: cannot perform the following tasks:
- Run configure hook of "lxd" snap if present (run hook "configure": 
-----
cat: /proc/self/attr/current: Permission denied
/snap/lxd/x1/snap/hooks/configure: 6: exec: aa-exec: Permission denied
-----)
root@u1:~# snap install lxd --channel 5.0/edge
lxd (5.0/edge) git-b2227d7 from Canonical✓ installed
root@u1:~# snap install --dangerous /dev/shm/lxd_0+git.0f0540b4_amd64.snap 
lxd 0+git.0f0540b4 installed
```

So 18.04 with 4.15 `-virtual` kernel and ZFS 0.7 works as well.

**Comment by simondeziel** [comment]
The `nvidia-container` part fails to build:

```
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.001 :: + '[' x86_64 '!=' x86_64 ']'
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.001 :: + set -ex
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.001 :: + git config user.email noreply@lists.canonical.com
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.003 :: + git config user.name 'LXD snap builder'
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.004 :: + patch -p1
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.009 :: patching file Makefile
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.010 :: Hunk #2 succeeded at 165 (offset -3 lines).
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.010 :: + patch -p1
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.011 :: patching file src/driver.c
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.011 :: Hunk #1 succeeded at 123 (offset 1 line).
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.011 :: + mkdir -p /root/parts/nvidia-container/install/bin/
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.014 :: + cp /root/project/snapcraft/wrappers/nvidia-container-cli /root/parts/nvidia-container/install/bin/
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.028 :: + set +ex
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.600 Executing PosixPath('/root/parts/nvidia-container/run/build.sh')
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.602 :: + make -j6
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.630 :: Makefile:106: *** Invalid major version.  Stop.
2024-09-18 01:13:24.222 :: 2024-09-18 01:13:23.631 :: error:
```

**Comment by simondeziel** [comment]
The `nvidia-container` build issue seems to be because upstream had a problem with their release tagging. Both [v1.16.1](https://github.com/NVIDIA/libnvidia-container/releases/tag/v1.16.1) and [v1.16.0](https://github.com/NVIDIA/libnvidia-container/releases/tag/v1.16.0) point to the exact same commit https://github.com/NVIDIA/libnvidia-container/commit/4c2494f16573b585788a42e9c7bee76ecd48c73d

File: renovate.json [hunk 14]
```diff
@@ -0,0 +1,7 @@
+{
+  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
+  "extends": [
+    "config:base",
+    "schedule:earlyMondays"
```
Comment (tomponline) [inline renovate.json:5]: Can you align with https://github.com/canonical/lxd/blob/main/.github/renovate.json

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
@simondeziel does this include the fix for the snap preseed config?

**Comment by tomponline** [comment]
@simondeziel is this ready to merge?

**Comment by simondeziel** [comment]
> @simondeziel does this include the fix for the snap preseed config?

I did not include the breaking commit nor the one that reverted it. Bottom-line is the preseed/`FIRSTRUN` detection should work.

> @simondeziel is this ready to merge?

I finally got around to testing the resulting snap:

```
root@v2:~# lxd init --auto
root@v2:~# lxc launch images:alpine/edge a1
Creating a1
Starting a1                               
root@v2:~# lxc launch images:alpine/edge av1 --vm -c security.secureboot=false
Creating av1
Starting av1                                
root@v2:~# lxc ls
+------+---------+----------------------+-----------------------------------------------+-----------------+-----------+
| NAME |  STATE  |         IPV4         |                     IPV6                      |      TYPE       | SNAPSHOTS |
+------+---------+----------------------+-----------------------------------------------+-----------------+-----------+
| a1   | RUNNING | 10.172.22.219 (eth0) | fd42:7205:c1dc:a1e7:216:3eff:fe71:eafe (eth0) | CONTAINER       | 0         |
+------+---------+----------------------+-----------------------------------------------+-----------------+-----------+
| av1  | RUNNING | 10.172.22.26 (eth0)  | fd42:7205:c1dc:a1e7:216:3eff:fe8a:c961 (eth0) | VIRTUAL-MACHINE | 0         |
+------+---------+----------------------+-----------------------------------------------+-----------------+-----------+
root@v2:~# ll /snap/lxd/
total 8
drwxr-xr-x  4 root root 4096 Sep 20 20:41 ./
drwxr-xr-x  6 root root 4096 Sep 20 20:40 ../
drwxr-xr-x 21 root root  291 Sep 17 15:19 30317/
lrwxrwxrwx  1 root root    2 Sep 20 20:41 current -> x1/
drwxr-xr-x 19 root root  267 Sep 18 21:21 x1/
```
